src/lamplib/src/genny/tasks/compile.py

```diff
@@ -32,6 +33,85 @@ def _sanitizer_flags(sanitizer: str, genny_repo_root: str):
     raise ValueError("Unknown sanitizer {}".format(sanitizer))
 
 
+class DistroDetectionError(Exception):
+    pass
+
+
+def _detect_distro_ubuntu(machine, freedesktop_version):
+    if freedesktop_version == "22.04":
+        if machine == "aarch64":
+            return "ubuntu2204_arm64"
+        elif machine == "x86_64":
+            return "ubuntu2204"
+        else:
+            raise DistroDetectionError(f"Invalid machine type for Ubuntu 22.04: {machine}")
+    elif freedesktop_version == "20.04":
+        if machine == "aarch64":
+            return "ubuntu2004_arm64"
+        elif machine == "x86_64":
+            return "ubuntu2004"
+        else:
+            raise DistroDetectionError(f"Invalid machine type for Ubunutu 20.04: {machine}")
+    elif freedesktop_version == "18.04":
+        if machine == "x86_64":
+            return "ubuntu1804"
+        else:
+            raise DistroDetectionError(f"Invalid machine type for Ubuntu 18.04: {machine}")
+    else:
+        raise DistroDetectionError(f"Invalid version for Ubuntu: {freedesktop_version}")
+
+
+def _detect_distro_rhel(machine, freedesktop_version):
+    # We only distinguish between major versions of RHEL, but they include minor versions in their version_id
+    if freedesktop_version.startswith("7."):
+        if machine == "x86_64":
+            return "rhel70"
+        else:
+            raise DistroDetectionError(f"Invalid machine type for RHEL 7.x: {machine}")
+    elif freedesktop_version.startswith("8."):
+        if machine == "x86_64":
+            return "rhel8"
+        else:
+            raise DistroDetectionError(f"Invalid machine type for RHEL 8.x: {machine}")
+
+
+def _detect_distro_amazon(machine, freedesktop_version):
+    if freedesktop_version == "2":
+        if machine == "aarch64":
+            return "amazon2_arm64"
+        elif machine == "x86_64":
+            return "amazon2"
+        else:
+            raise DistroDetectionError(f"Invalid machine type for Amazon 2: {machine}")
+    else:
+        raise DistroDetectionError(f"Invalid version for Amazon Linux: {freedesktop_version}")
+
+
+def detect_distro():
+    SLOG.info(f"Distro not specified. Detecting distro.")
+
+    system = platform.system()
+    machine = platform.machine()
+    if system == "Darwin":
+        distro = "not-linux"
+    elif system == "Linux":
+        freedesktop_release = platform.freedesktop_os_release()
```

Comment:
A risk involved with using `platform.freedesktop_os_release()` is that [it's only available from Python 3.10](https://docs.python.org/3/library/platform.html#:~:text=New%20in%20version%203.10.), and we haven't decided what's the lowest version of Python that we want Genny to support. Evergreen uses Python 3.9 IIRC, but if the `-d` flag is set for `./run-genny install` then it would be the same as before, so should be okay for automation on Evergreen. However, for manual use on MacOS, Genny users might be used to not setting this flag, so if they're using a Python version before 3.10 then this would be a UX change for them. 

Maybe we can add a Python version check using [`platform.python_version_tuple()`](https://docs.python.org/3/library/platform.html#platform.python_version_tuple), and error out to tell the user to set the `-d` flag if they are using a Python version before 3.10?

Reply:
This is a good point. It might be a little more complicated to get the version without that, but hopefully not that much more complicated?

Reply:
I fixed this by adding our own `_freedesktop_os_release` function. Since it's relatively easy to test if all our distros are working, this seems not too crazy bad. I also tried to make clear in the comment the _why_ of it, so hopefully future code readers know they can get rid of it.